### PR TITLE
feat: data plane & envoyproxy resilience test suite

### PR DESCRIPTION
**Pull request by alexwo**
**What type of PR is this?**
This pr adds a resilience test suite to cover the expected behavior of the data plane for both existing and new proxy instances during specific edge conditions.

**Why do we use Cilium?**
The default network plugin in Kubernetes-in-Docker (Kind) does not support enforcing network policies that restrict connectivity to local services, such as the Kubernetes API server. However, by disabling the default CNI and deploying Cilium as the CNI plugin, we can implement network policies that effectively control access to the API server. This approach allows for precise management of network traffic within the cluster, including the ability to block connectivity to the Kubernetes API server as needed.

**New test suite**
Having a new resilience test suite can be helpful due to the unique network setup and tests involved in resilience testing, which differ significantly from those of existing test suites.

For instance, we need a custom CNI only for this test case, other tests can still use the default.

Fault Injection: Resilience testing involves deliberately introducing failures to assess the system’s response and recovery capabilities. This approach is not typically encompassed by standard functional or performance test suites.

By maintaining a separate resilience test suite, we can effectively simulate and manage system stressors, and separate this concern from the other test cases we have.

Test suite is expected to pass once https://github.com/envoyproxy/gateway/pull/4767 is merged.

Issue:
https://github.com/envoyproxy/gateway/issues/4861

### CODE REVIEW

File: test/resilience/tests/envoyproxy.go
Comment (zirain) [inline test/resilience/tests/envoyproxy.go:30]: why we need this here?

File: tools/hack/create-cluster.sh
Comment (zirain) [inline tools/hack/create-cluster.sh:147]: can you share more details about why we need cilium?

**Comment by zirain** [comment]
why we need a new test suite instead of reuse existing one?

File: test/resilience/tests/envoyproxy.go
Comment (alexwo) [reply test/resilience/tests/envoyproxy.go:30]: its Integrating the Kubernetes client (client-go) with the zap logging library enhances the visibility of internal operations and API interactions, facilitating more effective debugging and monitoring. so incase things fail, its helps to show some of internal client errors. however we can skip that.

File: tools/hack/create-cluster.sh
Comment (alexwo) [reply tools/hack/create-cluster.sh:147]: The default network plugin in Kubernetes-in-Docker (Kind) does not support enforcing network policies that restrict connectivity to local services, such as the Kubernetes API server. However, by disabling the default CNI and deploying Cilium as the CNI plugin, we can implement network policies that effectively control access to the API server. This approach allows for precise management of network traffic within the cluster, including the ability to block connectivity to the Kubernetes API server as needed.

**Comment by alexwo** [comment]
> Contributor

> why we need a new test suite instead of reuse existing one?
Having a dedicated resilience test suite is helpful due to the unique network setup and tests involved in resilience testing, which differ significantly from those of existing test suites. 

Fault Injection: Resilience testing involves deliberately introducing failures to assess the system’s response and recovery capabilities. This approach is not typically encompassed by standard functional or performance test suites.

By maintaining a separate resilience test suite, we can effectively simulate and manage system stressors, and separate this concern from the other test cases we have.

File: test/resilience/tests/envoyproxy.go
Comment (zirain) [reply test/resilience/tests/envoyproxy.go:30]: it should be in main or somewhere, better not to be in `init` like https://github.com/envoyproxy/gateway/blob/0f68219d191b2b1d5953d59945c5e174db229c1c/test/e2e/e2e_test.go#L30

File: tools/hack/create-cluster.sh
Comment (zirain) [reply tools/hack/create-cluster.sh:147]: I recall kind support network policy as https://github.com/kubernetes-sigs/kind/issues/842 closed.

File: test/resilience/tests/envoyproxy.go
Comment (alexwo) [reply test/resilience/tests/envoyproxy.go:30]: indeed moved that.

File: test/resilience/tests/envoyproxy.go
Comment (guydc) [inline test/resilience/tests/envoyproxy.go:67]: Can we use something similar to the conformance suite test util that these calls from outside the cluster and wait for x consecutive successes? Is there some limitation on using the metallb LB svc address when using cilium?

File: test/resilience/tests/envoyproxy.go
Comment (guydc) [inline test/resilience/tests/envoyproxy.go:68]: Is there some sort of envoy XDS cache invalidation (e.g. https://www.envoyproxy.io/docs/envoy/latest/api-docs/xds_protocol#ttl) that we should consider here to determine the time when requests should succeed before failure begin? 
@arkodg

File: test/resilience/tests/envoygateway.go
Comment (guydc) [inline test/resilience/tests/envoygateway.go:47]: nit: EnvoyGateway reconciles

File: test/resilience/tests/envoygateway.go
Comment (guydc) [inline test/resilience/tests/envoygateway.go:137]: Can we add an assertion that the new leader is actually doing some work besides being elected? e.g. add a new route and check that it's actually getting programmed on the proxies?

Comment (alexwo) [reply test/resilience/tests/envoygateway.go:137]: Sure, i will add it.

File: .github/workflows/build_and_test.yaml
Comment (alexwo) [reply .github/workflows/build_and_test.yaml:186]: not needed .. just a copy paste.

File: test/resilience/tests/envoygateway.go
Comment (alexwo) [reply test/resilience/tests/envoygateway.go:54]: In general we could use the lease resource to understand that there is a leader, however, we can't know the active leader pod. as the lease spec hold the deployment identify as holder "holderIdentity: envoy-gateway-control-plane"

As per:
`
spec:
  holderIdentity: envoy-gateway-control-plane
  leaseDurationSeconds: 40
  renewTime: "2024-12-09T22:19:14.325096Z"
`
For the above test its less interesting but we have a test that simulate a leader only connectivity to api server problem where this is more interesting to find the leading pod only.

We can add a metric to help us with that.

File: test/resilience/tests/envoyproxy.go
Comment (alexwo) [reply test/resilience/tests/envoyproxy.go:67]: Yes, this just made it more easy to run this suite from mac

File: tools/hack/create-cluster.sh
Comment (zirain) [reply tools/hack/create-cluster.sh:147]: will https://github.com/kubernetes-sigs/kube-network-policies help?

Comment (alexwo) [reply tools/hack/create-cluster.sh:147]: I will give it a try, maybe it can do the work :)

Comment (alexwo) [reply tools/hack/create-cluster.sh:147]: @zirain I tested with kube-network-policies, but it was unable to block connectivity to the API server. In fact, I tried all the other available options as well. Currently, it seems this functionality is only achievable with the Cilium CNI.

File: test/resilience/tests/envoygateway.go
Comment (guydc) [reply test/resilience/tests/envoygateway.go:54]: If all the pods think of themselves as having the same identity, how do they know that they are not the current leader when trying to acquire the lease? Is that based on some other meta field?

File: test/resilience/tests/envoyproxy.go
Comment (guydc) [reply test/resilience/tests/envoyproxy.go:67]: ah, yea, that's a general problem for all of our tests when developing on macos. So, I do prefer that we use the regular tooling.

File: test/resilience/tests/envoygateway.go
Comment (alexwo) [reply test/resilience/tests/envoygateway.go:54]: hmm actually, holderIdentity seems to hold the elected pod identity. Not sure why i saw it as  envoy-gateway-control-plane. we can use that.

File: test/resilience/tests/envoyproxy.go
Comment (alexwo) [reply test/resilience/tests/envoyproxy.go]: done

File: tools/hack/create-cluster.sh
Comment (zhaohuabing) [inline tools/hack/create-cluster.sh:55]: Nit, non-blocking: Calico also works for "normal" e2e tests?  We can simply remove the `CUSTOM_CNI` falg and install it by default?

File: test/resilience/tests/envoygateway.go
Comment (zhaohuabing) [inline test/resilience/tests/envoygateway.go:77]: Question: Even though the leader pod is done due to losing connection to the API server, the replicas will still reach 3 because the deployment has been scaled to 3 replicas?

File: test/resilience/tests/envoygateway.go
Comment (zhaohuabing) [inline test/resilience/tests/envoygateway.go:127]: typo? envoy-gatay

File: test/resilience/tests/envoygateway.go
Comment (zhaohuabing) [inline test/resilience/tests/envoygateway.go:80]: Can we apply a new `HTTPRoute` and test it?
This ensures that the xDS comes from the newly eleceted EG instance.

File: test/resilience/tests/envoygateway.go
Comment (zhaohuabing) [inline test/resilience/tests/envoygateway.go:81]: Can we create a new HTTPRoute and test it?
This ensures that the xDS comes from the newly eleceted EG instance.

**Comment by zhaohuabing** [comment]
Overall, it looks great! This test will definitely improve the stability of EG. It seems the three Envoy Gateway tests can be merged into one.

**Comment by zhaohuabing** [comment]
> I would prefer to make this as part of e2e if there's no special reason. Especailly don't want to introudce a whole new test suite framework.

@envoyproxy/gateway-maintainers please share your options.

It could be included in the "normal" test suite, but I don't have a strong opinion on this.

File: test/resilience/tests/envoygateway.go
Comment (alexwo) [reply test/resilience/tests/envoygateway.go:77]: You’re absolutely right to consider the intermediate state during the transition, but it’s important to clarify how Kubernetes handles pods that lose connectivity to the API server. When a leader pod loses connectivity to the API server, Kubernetes does not immediately terminate or stop the pod. Instead, the pod itself detects the loss of connectivity, initiates a graceful teardown process, and restarts to attempt to reconnect to the API server. 

The replica count for the deployment remains at 3 throughout the process.

Kubernetes does not schedule a new pod to replace the one that lost connectivity because the existing pod is not considered failed from Kubernetes’ perspective. It’s the responsibility of the application running inside the pod (e.g., the leader election logic) to handle reconnection attempts or restart itself.

File: test/resilience/tests/envoygateway.go
Comment (alexwo) [reply test/resilience/tests/envoygateway.go:81]: In this test, we simulate an API failure affecting all pods, preventing any pod from assuming the leader role. Consequently, a leader transition does not occur. However, if the xDS configuration for the standby pods was calculated correctly, we would expect the /welcome route to remain accessible after the Envoy proxy reconnects to the secondary nodes. At this is not the case, as xDS is not having correct state in the secondary eg pods.

File: tools/hack/create-cluster.sh
Comment (alexwo) [reply tools/hack/create-cluster.sh:55]: For standard e2e tests, the default kindnet CNI is typically sufficient and offers a simpler, more resource-efficient setup. Introducing Calico can be reserved for scenarios that specifically require its advanced networking capabilities. 
At this point the Calico deployment adds an extra 2-3 min for each create-cluster.

File: test/resilience/tests/envoygateway.go
Comment (zhaohuabing) [reply test/resilience/tests/envoygateway.go:77]: Thanks for the explanation!

File: tools/hack/create-cluster.sh
Comment (zhaohuabing) [reply tools/hack/create-cluster.sh:55]: Yeah, an extra 2-3 minutes will slow down the tests, which already take a long time to finish.

**Review by zhaohuabing — DISMISSED** [review]

File: test/resilience/tests/envoygateway.go
Comment (guydc) [inline test/resilience/tests/envoygateway.go:69]: If all pods loose connectivity, don't we expect that they would all fail eventually once becoming leaders?

Comment (alexwo) [reply test/resilience/tests/envoygateway.go:69]: Pods rely on connectivity to the API server to participate in leader election processes. Without this connectivity, they cannot become leaders. Pods should continue to operate using their last known good configuration (xDS) and serve traffic accordingly.

**Comment by zhaohuabing** [comment]
@alexwo Could you please add a few lines of the comments for each test? (Like you explained in the response of this PR) This will help other contributors understand why we need the test and how the test works. Thanks!

**Review by zhaohuabing — APPROVED** [review]

**Comment by zhaohuabing** [comment]
@alexwo I think this is not right:

```
Waiting for deployment "envoy-gateway": ready replicas 3/2
Waiting for deployment "envoy-gateway": ready replicas 3/2
Waiting for deployment "envoy-gateway": ready replicas 3/2
Deployment "envoy-gateway" reached 2 replicas as expected.
    helpers.go:609: 2024-12-11T21:54:16.694551571Z: Conditions matched expectations
```

The ready replicas should be 2 even though https://github.com/envoyproxy/gateway/pull/4767 is not merged. This step only tests the leader pod losing its status.

**Review by zhaohuabing — DISMISSED** [review]

File: test/resilience/tests/envoygateway.go
Comment (zhaohuabing) [inline test/resilience/tests/envoygateway.go:71]: This change will allow test Resilience test suite with #4767 pass.  See https://github.com/envoyproxy/gateway/pull/4904

Not sure the exact reason, it seems like a bug from the Calico CNI. 

```suggestion
				return suite.Kube().ManageEgress(ctx, apiServerIP, namespace, policyName, true, map[string]string{"app.kubernetes.io/name": "gateway-helm"})
```

Comment (alexwo) [reply test/resilience/tests/envoygateway.go]: hmm i have merged the suggestion,  the Resilience suite still does not pass. 

Did this pass  for you when running locally? It’s possible that delays in resource cleanup caused the tests to behave inconsistently?

Comment (alexwo) [reply test/resilience/tests/envoygateway.go]: Understood. Without this fix, the tests fail even with the xDS adjustments. However, this minor change ensures that everything functions as expected. right?

**Comment by arkodg** [comment]
https://github.com/envoyproxy/gateway/pull/4767 has been merged, that should unblock this PR now ?

**Review by zhaohuabing — APPROVED** [review]

File: test/resilience/tests/envoygateway.go
Comment (zhaohuabing) [reply test/resilience/tests/envoygateway.go]: @alexwo The Envoy wasn't able to connect to the standby EG instances with the egress rule even though only the API server IP was blocked. Adding a podselector solved that. 

This looks like a Calico bug to me.

**Comment by zhaohuabing** [comment]
> #4767 has been merged, that should unblock this PR now ?

@arkodg I believe this PR is ready to be merged.

Should we also cherry-pick this to v1.2.4?

**Review by zhaohuabing — APPROVED** [review]

**Review by guydc — APPROVED** [review]

**Pull request merged**
